# **Best Model Selection in Pipeline**

To select the best model when using multiple models in a pipeline, you can use techniques like `cross-validation` and `evaluation metrics` to compare their performance. Here's an example of how to accomplish this on the Titanic dataset:

In [1]:
import pandas as pd
import seaborn as sns
from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from xgboost import XGBClassifier
from sklearn.metrics import accuracy_score

# Load the Titanic dataset from Seaborn
titanic_data = sns.load_dataset('titanic')

# Select features and target variable
X = titanic_data[['pclass', 'sex', 'age', 'fare', 'embarked']]
y = titanic_data['survived']

# Split the data into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Create a list of models to evaluate
models = [
    ('Random Forest', RandomForestClassifier(random_state=42)),
    ('Gradient Boosting', GradientBoostingClassifier(random_state=42)),
    ('XGBoost', XGBClassifier(random_state=42))
]

best_model = None
best_accuracy = 0.0

# Iterate over the models and evaluate their performance
for name, model in models:
    # Create a pipeline for each model
    pipeline = Pipeline([
        ('imputer', SimpleImputer(strategy='most_frequent')),
        ('encoder', OneHotEncoder(handle_unknown='ignore')),
        ('model', model)
    ])
    
    # Perform cross-validation
    scores = cross_val_score(pipeline, X_train, y_train, cv=5)
    
    # Calculate mean accuracy
    mean_accuracy = scores.mean()
    
    # Fit the pipeline on the training data
    pipeline.fit(X_train, y_train)
    
    # Make predictions on the test data
    y_pred = pipeline.predict(X_test)
    
    # Calculate accuracy score
    accuracy = accuracy_score(y_test, y_pred)
    
    # Print the performance metrics
    print("Model:", name)
    print("Cross-validation Accuracy:", mean_accuracy)
    print("Test Accuracy:", accuracy)
    print()
    
    # Check if the current model has the best accuracy
    if accuracy > best_accuracy:
        best_accuracy = accuracy
        best_model = pipeline

# Retrieve the best model
print("Best Model:", best_model)

Model: Random Forest
Cross-validation Accuracy: 0.7991529597163399
Test Accuracy: 0.8379888268156425

Model: Gradient Boosting
Cross-validation Accuracy: 0.8061952132374668
Test Accuracy: 0.7988826815642458

Model: XGBoost
Cross-validation Accuracy: 0.8076233625529401
Test Accuracy: 0.7932960893854749

Best Model: Pipeline(steps=[('imputer', SimpleImputer(strategy='most_frequent')),
                ('encoder', OneHotEncoder(handle_unknown='ignore')),
                ('model', RandomForestClassifier(random_state=42))])


The above code is using a `machine learning pipeline` to preprocess the Titanic dataset, `evaluate three different models` (RandomForestClassifier, GradientBoostingClassifier, and XGBoostClassifier), and `select the best model` based on `test accuracy`. Here's a breakdown of the code:

- The Titanic dataset is loaded and the features (`X`) and target variable (`y`) are selected.

- The data is split into training and test sets using `train_test_split`.

- A list of models to evaluate is defined. Each model is represented as a tuple containing the name of the model and the model itself.

- The code then iterates over the list of models. For each model, it creates a pipeline that first imputes missing values with the most frequent value, applies one-hot encoding, and then trains the model.

- The pipeline is evaluated using 5-fold cross-validation on the training data. The mean cross-validation accuracy is calculated and printed out.

- The pipeline is then fitted on the entire training data and used to make predictions on the test data. The test accuracy is calculated and printed out.

- If the test accuracy of the current model is higher than the best accuracy seen so far, the current model is updated as the best model.

- After all models have been evaluated, the best model is printed out.

**Observations from the output:**
> The output shows that the `RandomForestClassifier` has the highest test accuracy of approximately `83.80%`, so it is selected as the best model. This shows that even though other models might have higher cross-validation accuracy (like the GradientBoostingClassifier and XGBoostClassifier), the RandomForestClassifier performs best on the unseen test data.

---

## **Adding more Models in the above Code**

In [2]:
import pandas as pd
import seaborn as sns
from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Load the Titanic dataset from Seaborn
titanic_data = sns.load_dataset('titanic')

# Select features and target variable
X = titanic_data[['pclass', 'sex', 'age', 'fare', 'embarked']]
y = titanic_data['survived']

# Split the data into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Create a list of models to evaluate
models = [
    ('Random Forest', RandomForestClassifier(random_state=42)),
    ('Gradient Boosting', GradientBoostingClassifier(random_state=42)),
    ('Support Vector Machine', SVC(random_state=42)),
    ('Logistic Regression', LogisticRegression(random_state=42))
]

best_model = None
best_accuracy = 0.0

# Iterate over the models and evaluate their performance
for name, model in models:
    # Create a pipeline for each model
    pipeline = Pipeline([
        ('imputer', SimpleImputer(strategy='most_frequent')),
        ('encoder', OneHotEncoder(handle_unknown='ignore')),
        ('model', model)
    ])
    
    # Perform cross-validation
    scores = cross_val_score(pipeline, X_train, y_train, cv=5)
    
    # Calculate mean accuracy
    mean_accuracy = scores.mean()
    
    # Fit the pipeline on the training data
    pipeline.fit(X_train, y_train)
    
    # Make predictions on the test data
    y_pred = pipeline.predict(X_test)
    
    # Calculate accuracy score
    accuracy = accuracy_score(y_test, y_pred)
    
    # Print the performance metrics
    print("Model:", name)
    print("Cross-validation Accuracy:", mean_accuracy)
    print("Test Accuracy:", accuracy)
    print()
    
    # Check if the current model has the best accuracy
    if accuracy > best_accuracy:
        best_accuracy = accuracy
        best_model = pipeline

# Retrieve the best model
print("Best Model:", best_model)

Model: Random Forest
Cross-validation Accuracy: 0.7991529597163399
Test Accuracy: 0.8379888268156425

Model: Gradient Boosting
Cross-validation Accuracy: 0.8061952132374668
Test Accuracy: 0.7988826815642458

Model: Support Vector Machine
Cross-validation Accuracy: 0.8160248202501723
Test Accuracy: 0.8044692737430168

Model: Logistic Regression
Cross-validation Accuracy: 0.7977839062346105
Test Accuracy: 0.8100558659217877

Best Model: Pipeline(steps=[('imputer', SimpleImputer(strategy='most_frequent')),
                ('encoder', OneHotEncoder(handle_unknown='ignore')),
                ('model', RandomForestClassifier(random_state=42))])


The above code is using a `machine learning pipeline` to preprocess the Titanic dataset, `evaluate four different models` (RandomForestClassifier, GradientBoostingClassifier, Support Vector Machine, and Logistic Regression), and `select the best model` based on `test accuracy`. Here's a breakdown of the code:

- The Titanic dataset is loaded and the features (`X`) and target variable (`y`) are selected.
- The data is split into training and test sets using `train_test_split`.
- A list of models to evaluate is defined. Each model is represented as a tuple containing the name of the model and the model itself.
- The code then iterates over the list of models. For each model, it creates a pipeline that first imputes missing values with the most frequent value, applies one-hot encoding, and then trains the model.
- The pipeline is evaluated using 5-fold cross-validation on the training data. The mean cross-validation accuracy is calculated and printed out.
- The pipeline is then fitted on the entire training data and used to make predictions on the test data. The test accuracy is calculated and printed out.
- If the test accuracy of the current model is higher than the best accuracy seen so far, the current model is updated as the best model.
- After all models have been evaluated, the best model is printed out.

**Observations from the output:**
> The output shows that the `RandomForestClassifier` has the highest test accuracy of approximately `83.80%`, so it is selected as the best model. This shows that even though other models might have higher cross-validation accuracy (like the Support Vector Machine), the RandomForestClassifier performs best on the unseen test data.